turftopic/hierarchical.py

```diff
@@ -0,0 +1,277 @@
+    fig = fig.update_traces(
+        customdata=[[table] for table in tables],
+        hovertemplate="<b>%{text}</b> <br> <br> %{customdata[0]}",
+    )
+    fig = fig.update_traces(marker=dict(size=20, color="rgba(0,0,0,0.2)"))
+    fig = fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
+    fig = fig.update_yaxes(showgrid=False, visible=False, zeroline=False)
+    fig = fig.update_xaxes(
+        showgrid=False,
+        visible=False,
+        zeroline=False,
+        range=(xmin - xpad, xmax + xpad),
+    )
+    for start, end in edges_idx:
+        fig.add_shape(
+            type="line",
+            xref="x",
+            yref="y",
+            x0=x[start],
+            y0=y[start],
+            x1=x[end],
+            y1=y[end],
+            opacity=0.2,
+        )
+    return fig
+
+
+@dataclass
+class TopicNode:
+    """Node for a topic in a topic hierarchy.
+
+    Parameters
+    ----------
+    model: ContextualModel
+        Underlying topic model, which the hierarchy is based on.
+    path: tuple[int], default ()
+        Path that leads to this node from the root of the tree.
+    word_importance: ndarray of shape (n_vocab), default None
+        Importance of each word in the vocabulary for given topic.
+    document_topic_vector: ndarray of shape (n_documents), default None
+        Importance of the topic in all documents in the corpus.
+    children: list[TopicNode], default None
+        List of subtopics within this topic.
+    """
+
+    model: ContextualModel
+    path: tuple[int] = ()
+    word_importance: Optional[np.ndarray] = None
+    document_topic_vector: Optional[np.ndarray] = None
+    children: Optional[list[TopicNode]] = None
+
+    @classmethod
+    def create_root(
+        cls,
+        model: ContextualModel,
+        components: np.ndarray,
+        document_topic_matrix: np.ndarray,
+    ) -> TopicNode:
+        """Creates root node from a topic models' components and topic importances in documents."""
+        children = []
+        n_components = components.shape[0]
+        for i, comp, doc_top in zip(
+            range(n_components), components, document_topic_matrix.T
+        ):
+            children.append(
+                cls(
+                    model,
+                    path=(i,),
+                    word_importance=comp,
+                    document_topic_vector=doc_top,
+                    children=None,
+                )
+            )
+        return TopicNode(
+            model,
+            path=(),
+            word_importance=None,
+            document_topic_vector=None,
+            children=children,
+        )
+
+    @property
+    def level(self) -> int:
+        """Indicates how deep down the hierarchy the topic is."""
+        return len(self.path)
+
+    def get_words(self, top_k: int = 10) -> list[tuple[str, float]]:
+        """Returns top words and words importances for the topic.
+
+        Parameters
+        ----------
+        top_k: int, default 10
```

Comment:
I would generally not add the type information in both the docstring and the type hint (most doc renderers will fill it out automatically)

Reply:
It can sometime lead to issue where one is changed but the other one isn't - which one is then true?

Reply:
I know, it's a bad habit. I really like the NumPy dosctring style, and I wrote most of the library in it.
I think this would be something for a future PR to fix.